petastorm/tests/test_spark_dataset_converter.py

```diff
@@ -42,25 +44,48 @@ def test_primitive(self):
+                    self.assertEqual(getattr(ts, col)[i], expected_df[i][col])
 
-            self.assertEquals(len(converter), len(expected_df))
+            self.assertEqual(len(converter), len(expected_df))
 
-        self.assertEquals(ts.bool_col.dtype.type, np.bool_, "Boolean type column is not inferred correctly.")
-        self.assertEquals(ts.float_col.dtype.type, np.float32, "Float type column is not inferred correctly.")
-        self.assertEquals(ts.double_col.dtype.type, np.float64, "Double type column is not inferred correctly.")
-        self.assertEquals(ts.short_col.dtype.type, np.int16, "Short type column is not inferred correctly.")
-        self.assertEquals(ts.int_col.dtype.type, np.int32, "Integer type column is not inferred correctly.")
-        self.assertEquals(ts.long_col.dtype.type, np.int64, "Long type column is not inferred correctly.")
+        self.assertEqual(ts.bool_col.dtype.type, np.bool_, "Boolean type column is not inferred correctly.")
+        self.assertEqual(ts.float_col.dtype.type, np.float32, "Float type column is not inferred correctly.")
+        self.assertEqual(ts.double_col.dtype.type, np.float64, "Double type column is not inferred correctly.")
+        self.assertEqual(ts.short_col.dtype.type, np.int16, "Short type column is not inferred correctly.")
+        self.assertEqual(ts.int_col.dtype.type, np.int32, "Integer type column is not inferred correctly.")
+        self.assertEqual(ts.long_col.dtype.type, np.int64, "Long type column is not inferred correctly.")
 
     def test_delete(self):
         test_path = "/tmp/petastorm_test"
-        os.mkdir(test_path)
-        os.mkdir(os.path.join(test_path, "dir1"))
+        Path(test_path).mkdir(parents=True, exist_ok=True)
+        Path(os.path.join(test_path, "dir1")).mkdir(parents=True, exist_ok=True)
         with open(os.path.join(test_path, "file1"), "w") as f:
             f.write("abc")
         with open(os.path.join(test_path, "file2"), "w") as f:
             f.write("123")
         converter = SparkDatasetConverter(test_path, 0)
         converter.delete()
         self.assertFalse(os.path.exists(test_path))
+
+    def test_atexit(self):
+        cache_dir = "/tmp/123"
+        Path(cache_dir).mkdir(parents=True, exist_ok=True)
+        lines = """
+            from petastorm.spark.spark_dataset_converter import make_spark_converter
+            from pyspark.sql import SparkSession
+            import os
+            spark = SparkSession.builder.getOrCreate()
+            df = spark.createDataFrame([(1, 2),(4, 5)], ["col1", "col2"])
+            converter = make_spark_converter(df, '/tmp/123')
+            assert(os.path.exists(converter.cache_file_path))
+            f = open("/tmp/123/output", "w")
+            f.write(converter.cache_file_path)
+            f.close()
+            """
+        code_str = "; ".join(line.strip() for line in lines.strip().splitlines())
```

Comment:
We can improve this by using `multiprocessing.Process`, which could accept a python function as process main function.
But it is ok for now.

Reply:
I found some discussion about multiprocessing vs subprocess: https://stackoverflow.com/questions/13606867/what-is-the-difference-between-multiprocessing-and-subprocess IMO, this test is more about testing an independent python process than doing things faster by running code in parallel.

Reply:
Yes. But Multiprocessing also spawns sub process, and it can avoid write subprocess code in python string, which is beneficial to code maintenance. We can discuss and improve it in later PR.